File: server/src/api/v1/services/hotel.service.js
// src/api/v1/services/hotel.service.js
const hotelRepository = require('../repositories/hotel.repository');
const { AppError } = require('../../../utils/errors');

class HotelService {
  /**
   * Tạo mới khách sạn
   * @param {Object} hotelData - Dữ liệu khách sạn
   * @param {string} ownerId - ID chủ sở hữu
   * @returns {Promise<Object>}
   */
  async createHotel(hotelData, ownerId) {
    try {
      const hotel = await hotelRepository.create(hotelData, ownerId);

      return {
        success: true,
        message: 'Khách sạn đã được tạo thành công và đang chờ xét duyệt',
        data: hotel
      };
    } catch (error) {
      console.error('Lỗi khi tạo khách sạn:', error);
      if (error.code === '23505') {
        throw new AppError('Khách sạn với thông tin này đã tồn tại', 409);
      }
      throw error;
    }
  }

  /**
   * Lấy thông tin khách sạn theo ID
   */
  async getHotelById(hotelId) {
    const hotel = await hotelRepository.findById(hotelId);

    if (!hotel) {
      throw new AppError('Không tìm thấy khách sạn', 404);
    }

    return {
      success: true,
      data: hotel
    };
  }

  /**
   * Lấy danh sách tất cả khách sạn (có phân trang)
   */
  async getAllHotels(options = {}) {
    const { page = 1, limit = 10 } = options;
    const offset = (page - 1) * limit;

    const hotels = await hotelRepository.findAll(limit, offset);

    return {
      success: true,
      data: hotels,
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total: hotels.length
      }
    };
  }

  /**
   * Cập nhật khách sạn
   */
  async updateHotel(hotelId, updateData, userId) {
    const existingHotel = await hotelRepository.findById(hotelId);
    if (!existingHotel) {
      throw new AppError('Không tìm thấy khách sạn', 404);
    }

    if (existingHotel.ownerId !== userId) {
      throw new AppError('Bạn không có quyền cập nhật khách sạn này', 403);
    }

    const updatedHotel = await hotelRepository.update(hotelId, updateData);

    return {
      success: true,
      message: 'Cập nhật khách sạn thành công',
      data: updatedHotel
    };
  }

  /**
   * Xóa khách sạn (soft delete)
   */
  async deleteHotel(hotelId, userId) {
    const existingHotel = await hotelRepository.findById(hotelId);
    if (!existingHotel) {
      throw new AppError('Không tìm thấy khách sạn', 404);
    }

    if (existingHotel.ownerId !== userId) {
      throw new AppError('Bạn không có quyền xóa khách sạn này', 403);
    }

    await hotelRepository.softDelete(hotelId);

    return {
      success: true,
      message: 'Xóa khách sạn thành công'
    };
  }

  /**
   * Lấy khách sạn theo trạng thái
   */
  async getHotelsByStatus(status) {
    const validStatuses = ['pending', 'approved', 'rejected', 'active', 'inactive'];
    if (!validStatuses.includes(status)) {
      throw new AppError('Trạng thái không hợp lệ', 400);
    }

    const hotels = await hotelRepository.findByStatus(status);

    return {
      success: true,
      data: hotels
    };
  }

  /**
   * Cập nhật trạng thái khách sạn (Admin)
   */
  async updateHotelStatus(hotelId, newStatus) {
    const validStatuses = ['pending', 'approved', 'rejected'];
    if (!validStatuses.includes(newStatus)) {
      throw new AppError('Trạng thái không hợp lệ', 400);
    }

    const existingHotel = await hotelRepository.findById(hotelId);
    if (!existingHotel) {
      throw new AppError('Không tìm thấy khách sạn', 404);
    }

    const updatedHotel = await hotelRepository.updateStatus(hotelId, newStatus);

    return {
      success: true,
      message: `Trạng thái khách sạn đã được cập nhật thành ${newStatus}`,
      data: updatedHotel
    };
  }

  /**
   * Lấy khách sạn theo chủ sở hữu
   */
  async getHotelsByOwner(ownerId) {
    const hotels = await hotelRepository.findByOwner(ownerId);

    return {
      success: true,
      data: hotels
    };
  }

  /**
   * Tìm kiếm khách sạn
   */
  async searchHotels(searchParams) {
    const { city, name, minRating, maxRating, page = 1, limit = 10 } = searchParams;
    const offset = (page - 1) * limit;

    let hotels = [];

    if (city) {
      hotels = await hotelRepository.findByCity(city);
    } else if (name) {
      hotels = await hotelRepository.searchByName(name);
    } else if (minRating || maxRating) {
      const min = minRating || 1;
      const max = maxRating || 5;
      hotels = await hotelRepository.findByRating(min, max);
    } else {
      const filters = {
        city,
        searchTerm: name,
        minRating,
        maxRating,
        status: 'approved'
      };
      hotels = await hotelRepository.findWithFilters(filters, limit, offset);
    }

    return {
      success: true,
      data: hotels,
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total: hotels.length
      }
    };
  }

  /**
   * Thống kê khách sạn
   */
  async getHotelStatistics() {
    const pendingCount = await hotelRepository.countByStatus('pending');
    const approvedCount = await hotelRepository.countByStatus('approved');
    const rejectedCount = await hotelRepository.countByStatus('rejected');
    const deletedCount = await hotelRepository.countByStatus('deleted');

    return {
      success: true,
      data: {
        pending: pendingCount,
        approved: approvedCount,
        rejected: rejectedCount,
        deleted: deletedCount,
        total: pendingCount + approvedCount + rejectedCount + deletedCount
      }
    };
  }

  /**
   * Lấy khách sạn phổ biến
   */
  async getPopularHotels(limit = 10) {
    const hotels = await hotelRepository.findByRating(4, 5);
    return {
      success: true,
      data: hotels.slice(0, limit)
    };
  }

  /**
   * Tìm kiếm khách sạn theo thành phố và phường/xã
   * @param {string} city - Tên thành phố/tỉnh
   * @param {string} ward - Tên phường/xã
   * @param {Object} options - Tùy chọn phân trang
   * @returns {Promise<Object>}
   */
  async getHotelsByCityAndWard(city, ward, options = {}) {
    const { page = 1, limit = 10 } = options;
    const offset = (page - 1) * limit;

    if (!city || !ward) {
      throw new AppError('Vui lòng cung cấp đầy đủ thông tin thành phố và phường/xã', 400);
    }

    const hotels = await hotelRepository.findByCityAndWard(city, ward, limit, offset);
    const totalCount = await hotelRepository.countByCityAndWard(city, ward);

    return {
      success: true,
      message: `Tìm thấy ${totalCount} khách sạn tại ${ward}, ${city}`,
      data: hotels,
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total: totalCount,
        totalPages: Math.ceil(totalCount / limit)
      }
    };
  }

  /**
   * Đếm số lượng khách sạn theo thành phố và phường/xã
   * @param {string} city - Tên thành phố/tỉnh
   * @param {string} ward - Tên phường/xã
   * @returns {Promise<Object>}
   */
  async countHotelsByCityAndWard(city, ward) {
    if (!city || !ward) {
      throw new AppError('Vui lòng cung cấp đầy đủ thông tin thành phố và phường/xã', 400);
    }

    const count = await hotelRepository.countByCityAndWard(city, ward);

    return {
      success: true,
      message: `Số lượng khách sạn tại ${ward}, ${city}`,
      data: {
        city,
        ward,
        count
      }
    };
  }
}

module.exports = new HotelService();

File: mobile/lib/services/hotel_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../classes/hotel_model.dart';
import 'api_config.dart';
import 'token_service.dart';

class HotelService {
  // Singleton pattern
  static final HotelService _instance = HotelService._internal();
  factory HotelService() => _instance;
  HotelService._internal();

  // Headers mặc định
  Map<String, String> get _headers => {
    'Content-Type': 'application/json',
    'Accept': 'application/json',
  };

  // Headers có token
  Map<String, String> _headersWithToken(String token) => {
    'Content-Type': 'application/json',
    'Accept': 'application/json',
    'Authorization': 'Bearer $token',
  };

  // ============================================
  // PUBLIC METHODS
  // ============================================

  /// Lấy danh sách tất cả khách sạn (có phân trang)
  /// GET /api/v1/hotels
  Future<Map<String, dynamic>> getAllHotels({
    int page = 1,
    int limit = 10,
  }) async {
    try {
      final url = Uri.parse(
        '${ApiConfig.baseUrl}/hotels?page=$page&limit=$limit',
      );

      final response = await http.get(url, headers: _headers);
      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        List<Hotel> hotels = [];
        if (responseData['data'] != null) {
          hotels = (responseData['data'] as List)
              .map((json) => Hotel.fromJson(json))
              .toList();
        }

        return {
          'success': true,
          'message':
              responseData['message'] ?? 'Lấy danh sách khách sạn thành công',
          'data': hotels,
          'pagination': responseData['pagination'],
        };
      } else {
        return {
          'success': false,
          'message':
              responseData['message'] ?? 'Lỗi khi lấy danh sách khách sạn',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Lấy thông tin khách sạn theo ID
  /// GET /api/v1/hotels/:id
  Future<Map<String, dynamic>> getHotelById(String hotelId) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/hotels/$hotelId');

      final response = await http.get(url, headers: _headers);
      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        Hotel? hotel;
        if (responseData['data'] != null) {
          hotel = Hotel.fromJson(responseData['data']);
        }

        return {
          'success': true,
          'message':
              responseData['message'] ?? 'Lấy thông tin khách sạn thành công',
          'data': hotel,
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Không tìm thấy khách sạn',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Tìm kiếm khách sạn
  /// GET /api/v1/hotels/search
  Future<Map<String, dynamic>> searchHotels({
    String? city,
    String? name,
    double? minRating,
    double? maxRating,
    int page = 1,
    int limit = 10,
  }) async {
    try {
      String queryString = 'page=$page&limit=$limit';

      if (city != null && city.isNotEmpty) {
        queryString += '&city=${Uri.encodeComponent(city)}';
      }
      if (name != null && name.isNotEmpty) {
        queryString += '&name=${Uri.encodeComponent(name)}';
      }
      if (minRating != null) {
        queryString += '&minRating=$minRating';
      }
      if (maxRating != null) {
        queryString += '&maxRating=$maxRating';
      }

      final url = Uri.parse('${ApiConfig.baseUrl}/hotels/search?$queryString');

      final response = await http.get(url, headers: _headers);
      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        List<Hotel> hotels = [];
        if (responseData['data'] != null) {
          hotels = (responseData['data'] as List)
              .map((json) => Hotel.fromJson(json))
              .toList();
        }

        return {
          'success': true,
          'message': responseData['message'] ?? 'Tìm kiếm thành công',
          'data': hotels,
          'pagination': responseData['pagination'],
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Lỗi khi tìm kiếm khách sạn',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Lấy khách sạn phổ biến
  /// GET /api/v1/hotels/popular
  Future<Map<String, dynamic>> getPopularHotels({int limit = 10}) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/hotels/popular?limit=$limit');

      final response = await http.get(url, headers: _headers);
      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        List<Hotel> hotels = [];
        if (responseData['data'] != null) {
          hotels = (responseData['data'] as List)
              .map((json) => Hotel.fromJson(json))
              .toList();
        }

        return {
          'success': true,
          'message':
              responseData['message'] ??
              'Lấy danh sách khách sạn phổ biến thành công',
          'data': hotels,
        };
      } else {
        return {
          'success': false,
          'message':
              responseData['message'] ??
              'Lỗi khi lấy danh sách khách sạn phổ biến',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  // ============================================
  // AUTHENTICATED METHODS
  // ============================================

  /// Lấy khách sạn của user hiện tại
  /// GET /api/v1/hotels/my-hotels
  Future<Map<String, dynamic>> getMyHotels(String token) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/hotels/my-hotels');

      final response = await http.get(url, headers: _headersWithToken(token));
      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        List<Hotel> hotels = [];
        if (responseData['data'] != null) {
          hotels = (responseData['data'] as List)
              .map((json) => Hotel.fromJson(json))
              .toList();
        }

        return {
          'success': true,
          'message':
              responseData['message'] ??
              'Lấy danh sách khách sạn của bạn thành công',
          'data': hotels,
        };
      } else {
        return {
          'success': false,
          'message':
              responseData['message'] ?? 'Lỗi khi lấy danh sách khách sạn',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Lấy khách sạn theo chủ sở hữu
  /// GET /api/v1/hotels/owner/:ownerId
  Future<Map<String, dynamic>> getHotelsByOwner(
    String ownerId,
    String token,
  ) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/hotels/owner/$ownerId');

      final response = await http.get(url, headers: _headersWithToken(token));
      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        List<Hotel> hotels = [];
        if (responseData['data'] != null) {
          hotels = (responseData['data'] as List)
              .map((json) => Hotel.fromJson(json))
              .toList();
        }

        return {
          'success': true,
          'message':
              responseData['message'] ?? 'Lấy danh sách khách sạn thành công',
          'data': hotels,
        };
      } else {
        return {
          'success': false,
          'message':
              responseData['message'] ?? 'Lỗi khi lấy danh sách khách sạn',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Tạo mới khách sạn
  /// POST /api/v1/hotels
  Future<Map<String, dynamic>> createHotel(
    Map<String, dynamic> hotelData,
    String token,
  ) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/hotels');

      final response = await http.post(
        url,
        headers: _headersWithToken(token),
        body: jsonEncode(hotelData),
      );

      final responseData = jsonDecode(response.body);

      if (response.statusCode == 201) {
        Hotel? hotel;
        if (responseData['data'] != null) {
          hotel = Hotel.fromJson(responseData['data']);
        }

        return {
          'success': true,
          'message':
              responseData['message'] ?? 'Khách sạn đã được tạo thành công',
          'data': hotel,
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Lỗi khi tạo khách sạn',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Cập nhật thông tin khách sạn
  /// PUT /api/v1/hotels/:id
  Future<Map<String, dynamic>> updateHotel(
    String hotelId,
    Map<String, dynamic> hotelData,
    String token,
  ) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/hotels/$hotelId');

      final response = await http.put(
        url,
        headers: _headersWithToken(token),
        body: jsonEncode(hotelData),
      );

      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        Hotel? hotel;
        if (responseData['data'] != null) {
          hotel = Hotel.fromJson(responseData['data']);
        }

        return {
          'success': true,
          'message': responseData['message'] ?? 'Cập nhật khách sạn thành công',
          'data': hotel,
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Lỗi khi cập nhật khách sạn',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Xóa khách sạn
  /// DELETE /api/v1/hotels/:id
  Future<Map<String, dynamic>> deleteHotel(String hotelId, String token) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/hotels/$hotelId');

      final response = await http.delete(
        url,
        headers: _headersWithToken(token),
      );
      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message': responseData['message'] ?? 'Xóa khách sạn thành công',
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Lỗi khi xóa khách sạn',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  // ============================================
  // ADMIN METHODS
  // ============================================

  /// Lấy tất cả khách sạn (Admin)
  /// GET /api/v1/hotels/admin/all
  Future<Map<String, dynamic>> getAllHotelsAdmin(
    String token, {
    int page = 1,
    int limit = 10,
    String? status,
  }) async {
    try {
      String queryString = 'page=$page&limit=$limit';
      if (status != null && status.isNotEmpty) {
        queryString += '&status=$status';
      }

      final url = Uri.parse(
        '${ApiConfig.baseUrl}/hotels/admin/all?$queryString',
      );

      final response = await http.get(url, headers: _headersWithToken(token));
      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        List<Hotel> hotels = [];
        if (responseData['data'] != null) {
          hotels = (responseData['data'] as List)
              .map((json) => Hotel.fromJson(json))
              .toList();
        }

        return {
          'success': true,
          'message':
              responseData['message'] ?? 'Lấy danh sách khách sạn thành công',
          'data': hotels,
          'pagination': responseData['pagination'],
        };
      } else {
        return {
          'success': false,
          'message':
              responseData['message'] ?? 'Lỗi khi lấy danh sách khách sạn',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Lấy khách sạn chờ duyệt (Admin)
  /// GET /api/v1/hotels/admin/pending
  Future<Map<String, dynamic>> getPendingHotels(String token) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/hotels/admin/pending');

      final response = await http.get(url, headers: _headersWithToken(token));
      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        List<Hotel> hotels = [];
        if (responseData['data'] != null) {
          hotels = (responseData['data'] as List)
              .map((json) => Hotel.fromJson(json))
              .toList();
        }

        return {
          'success': true,
          'message':
              responseData['message'] ??
              'Lấy danh sách khách sạn chờ duyệt thành công',
          'data': hotels,
        };
      } else {
        return {
          'success': false,
          'message':
              responseData['message'] ??
              'Lỗi khi lấy danh sách khách sạn chờ duyệt',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Lấy thống kê khách sạn (Admin)
  /// GET /api/v1/hotels/admin/statistics
  Future<Map<String, dynamic>> getHotelStatistics(String token) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/hotels/admin/statistics');

      final response = await http.get(url, headers: _headersWithToken(token));
      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message':
              responseData['message'] ?? 'Lấy thống kê khách sạn thành công',
          'data': responseData['data'],
        };
      } else {
        return {
          'success': false,
          'message':
              responseData['message'] ?? 'Lỗi khi lấy thống kê khách sạn',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Cập nhật trạng thái khách sạn (Admin)
  /// PATCH /api/v1/hotels/admin/:id/status
  Future<Map<String, dynamic>> updateHotelStatus(
    String hotelId,
    String status,
    String token,
  ) async {
    try {
      final url = Uri.parse(
        '${ApiConfig.baseUrl}/hotels/admin/$hotelId/status',
      );

      final response = await http.patch(
        url,
        headers: _headersWithToken(token),
        body: jsonEncode({'status': status}),
      );

      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        Hotel? hotel;
        if (responseData['data'] != null) {
          hotel = Hotel.fromJson(responseData['data']);
        }

        return {
          'success': true,
          'message':
              responseData['message'] ??
              'Cập nhật trạng thái khách sạn thành công',
          'data': hotel,
        };
      } else {
        return {
          'success': false,
          'message':
              responseData['message'] ??
              'Lỗi khi cập nhật trạng thái khách sạn',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Lấy khách sạn theo trạng thái (Admin)
  /// GET /api/v1/hotels/admin/status/:status
  Future<Map<String, dynamic>> getHotelsByStatus(
    String status,
    String token,
  ) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/hotels/admin/status/$status');

      final response = await http.get(url, headers: _headersWithToken(token));
      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        List<Hotel> hotels = [];
        if (responseData['data'] != null) {
          hotels = (responseData['data'] as List)
              .map((json) => Hotel.fromJson(json))
              .toList();
        }

        return {
          'success': true,
          'message':
              responseData['message'] ??
              'Lấy danh sách khách sạn theo trạng thái thành công',
          'data': hotels,
        };
      } else {
        return {
          'success': false,
          'message':
              responseData['message'] ??
              'Lỗi khi lấy danh sách khách sạn theo trạng thái',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  // ============================================
  // HELPER METHODS
  // ============================================

  /// Phê duyệt khách sạn (Admin)
  Future<Map<String, dynamic>> approveHotel(String hotelId, String token) {
    return updateHotelStatus(hotelId, 'approved', token);
  }

  /// Từ chối khách sạn (Admin)
  Future<Map<String, dynamic>> rejectHotel(String hotelId, String token) {
    return updateHotelStatus(hotelId, 'rejected', token);
  }

  /// Kích hoạt khách sạn (Admin)
  Future<Map<String, dynamic>> activateHotel(String hotelId, String token) {
    return updateHotelStatus(hotelId, 'active', token);
  }

  /// Vô hiệu hóa khách sạn (Admin)
  Future<Map<String, dynamic>> deactivateHotel(String hotelId, String token) {
    return updateHotelStatus(hotelId, 'inactive', token);
  }

  /// Tìm kiếm khách sạn theo thành phố và phường
  /// GET /api/v1/hotels/search/location?city=...&ward=...
  Future<Map<String, dynamic>> searchHotelsByLocation({
    required String city,
    required String ward,
    int page = 1,
    int limit = 10,
  }) async {
    try {
      String queryString = 'page=$page&limit=$limit';
      queryString += '&city=${Uri.encodeComponent(city)}';
      queryString += '&ward=${Uri.encodeComponent(ward)}';

      final url = Uri.parse(
        '${ApiConfig.baseUrl}/hotels/search/location?$queryString',
      );

      final response = await http.get(url, headers: _headers);
      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        List<Hotel> hotels = [];
        if (responseData['data'] != null) {
          hotels = (responseData['data'] as List)
              .map((json) => Hotel.fromJson(json))
              .toList();
        }

        return {
          'success': true,
          'message':
              responseData['message'] ??
              'Tìm kiếm khách sạn theo vị trí thành công',
          'data': hotels,
          'pagination': responseData['pagination'],
        };
      } else {
        return {
          'success': false,
          'message':
              responseData['message'] ??
              'Lỗi khi tìm kiếm khách sạn theo vị trí',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Đếm số khách sạn theo thành phố và phường
  /// GET /api/v1/hotels/count/location?city=...&ward=...
  Future<Map<String, dynamic>> countHotelsByLocation({
    required String city,
    required String ward,
  }) async {
    try {
      String queryString =
          'city=${Uri.encodeComponent(city)}&ward=${Uri.encodeComponent(ward)}';

      final url = Uri.parse(
        '${ApiConfig.baseUrl}/hotels/count/location?$queryString',
      );

      final response = await http.get(url, headers: _headers);
      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message':
              responseData['message'] ?? 'Đếm khách sạn theo vị trí thành công',
          'data': responseData['data'],
          'count': responseData['count'] ?? 0,
        };
      } else {
        return {
          'success': false,
          'message':
              responseData['message'] ?? 'Lỗi khi đếm khách sạn theo vị trí',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Lấy danh sách tiện nghi của khách sạn
  /// GET /api/v1/hotels/:hotelId/amenities
  Future<Map<String, dynamic>> getAmenitiesForHotel(String hotelId) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/hotels/$hotelId/amenities');

      final response = await http.get(url, headers: _headers);
      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        // Nếu có model Amenity thì parse, còn không trả raw data
        final amenities = responseData['data'];
        return {
          'success': true,
          'message': responseData['message'] ?? 'Lấy danh sách tiện nghi thành công',
          'data': amenities,
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Lỗi khi lấy danh sách tiện nghi',
        };
      }
    } catch (e) {
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }

  /// Lấy tất cả hình ảnh của một khách sạn
  /// GET /api/v1/hotels/:hotelId/images
  Future<Map<String, dynamic>> getHotelImages(String hotelId) async {
    try {
      final url = Uri.parse('${ApiConfig.baseUrl}/hotels/$hotelId/images');
      print('Calling API: $url'); // Debug log
      
      // Lấy token từ TokenService
      final token = await TokenService.getToken();
      
      // Sử dụng headers có token nếu có, không thì dùng headers thường
      final headers = token != null ? _headersWithToken(token) : _headers;
      
      final response = await http.get(url, headers: headers);
      print('Response status: ${response.statusCode}'); // Debug log
      print('Response body: ${response.body}'); // Debug log
      
      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body);
        return {
          'success': true,
          'message': responseData['message'] ?? 'Lấy danh sách ảnh thành công',
          'data': responseData['data'],
        };
      } else {
        final responseData = jsonDecode(response.body);
        return {
          'success': false,
          'message': responseData['message'] ?? 'Lỗi khi lấy danh sách ảnh',
        };
      }
    } catch (e) {
      print('Exception in getHotelImages: $e'); // Debug log
      return {'success': false, 'message': 'Lỗi kết nối: $e'};
    }
  }
}
